designs/philosophy.md

```diff
@@ -0,0 +1,169 @@
+# webauthn-rs Design Philosophy
+
+* **Published:** 2023-02-xx
+* **Last updated:** 2023-02-xx
+
+This document describes the design philosophy of `webauthn-rs`, as well as some
+of the critical design decisions we've made over time.
+
+If you ever wondered *why* `webauthn-rs` does things a certain way, this is the
+document for you.
+
+## Introduction
+
+`webauthn-rs` is fairly opinionated about how one should use WebAuthn.
+
+The WebAuthn API that most developers use comes from the browser:
+
+* `navigator.credentials.create` to register
+* `navigator.credentials.get` to authenticate
+
+This API has many edge cases, and is easy to use in an insecure way, or in a way
+that can make your application inflexible.
+
+Our goal is that `webauthn-rs`:
+
+* has safe interfaces
+* encourages good design choices
+* is hard to use in an unsafe way
+
+This means we'll intentionally do some things differently to WebAuthn, or avoid
+some APIs.
+
+We have built an escape hatch: `webauthn-rs-core`. However, we don't make
+guarantees about how that works.
```

Comment:
I'd say "webauthn-rs-core more closely matches the webauthn spec and the possible lowlevel choices, but we do not guarantee that api's in this will remain stable".

Reply:
Done